### PR TITLE
Reimplemented Deposit and Withdrawal Extrinsics

### PR DESCRIPTION
**Pull request by zktony**
## Describe your changes
* Reimplemented Deposit and Withdrawal Extrinsics.
* Added XCM Support.
* Added test cases.
 
## Issue ticket number and link
[#4](https://github.com/Polkadex-Substrate/thea/issues/4)

### CODE REVIEW

**Comment by zktony** [comment]
@nakul1010, please review this PR. While resolving merge conflicts, I may have removed your code. 
|This PR requires @nakul1010 approval|

**Review by nakul1010 — DISMISSED** [review]
Didn't affect any of my prev code written. LGTM

**Comment by serhii-temchenko** [comment]
@zktony is it ready to be reviewed? If so - please add a label so I could see that it is ready.

File: pallets/asset-handler/Cargo.toml [hunk 1]
```diff
@@ -26,8 +26,13 @@ sp-core = { workspace = true, default-features = false }
 sp-std = { workspace = true, default-features = false }
 sp-runtime = { workspace = true, default-features = false }
 
+xcm = { workspace = true, default-features = false }
+
 chainbridge = {path = "../chainbridge", default-features = false}
 
+thea-primitives = {path = "../../thea-primitives", default-features = false}
```
Comment (serhii-temchenko) [inline pallets/asset-handler/Cargo.toml]: Please add places before and after `{}`

File: pallets/asset-handler/src/lib.rs [hunk 2]
```diff
@@ -344,12 +358,7 @@ pub mod pallet {
 
 			// Hash the resulting vector with Keccak256 Hashing Algorithm and retrieve first 16
 			// bytes
-			let derived_asset_id_hash = &keccak_256(derived_asset_id.as_ref())[0..16];
-
-			// Derive u128 from resulting bytes
-			let mut temp = [0u8; 16];
-			temp.copy_from_slice(derived_asset_id_hash);
-			let asset_id = u128::from_le_bytes(temp);
+			let asset_id = Self::get_asset_id(derived_asset_id);
```
Comment (serhii-temchenko) [inline pallets/asset-handler/src/lib.rs]: I'd moved this comment on the `get_asset_id` function directly.

File: pallets/asset-handler/src/lib.rs [hunk 3]
```diff
@@ -365,6 +374,31 @@ pub mod pallet {
 			Ok(())
 		}
 
+		/// Create Parachain Asset
+		#[pallet::weight(T::WeightInfo::create_asset(1))]
+		pub fn create_parachain_asset(
+			origin: OriginFor<T>,
+			asset: sp_std::boxed::Box<AssetId>,
+		) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/asset-handler/src/lib.rs:379]: Please add params into the dockblock.

File: pallets/thea-staking/Cargo.toml [hunk 4]
```diff
@@ -58,4 +58,4 @@ std = [
 runtime-benchmarks = [
 	"frame-benchmarking/runtime-benchmarks",
 ]
-try-runtime = ["frame-support/try-runtime"]
+try-runtime = ["frame-support/try-runtime"]
```
Comment (serhii-temchenko) [inline pallets/thea-staking/Cargo.toml]: Please add empty line at the end of file

File: pallets/thea/Cargo.toml [hunk 5]
```diff
@@ -28,12 +28,14 @@ sp-core = { workspace = true, default-features = false }
 sp-std = { workspace = true, default-features = false }
 sp-runtime = { workspace = true, default-features = false }
 
-chainbridge = { path = "../chainbridge", default-features = false }
-asset-handler = { path = "../asset-handler", default-features = false }
+chainbridge = {path = "../chainbridge", default-features = false}
+asset-handler = {path = "../asset-handler", default-features = false}
 polkadex-primitives = { workspace = true, default-features = false }
-thea-primitives = { path = "../../thea-primitives", default-features = false }
+thea-primitives = {path = "../../thea-primitives", default-features = false}
```
Comment (serhii-temchenko) [inline pallets/thea/Cargo.toml]: Please add spaces before and after `{}`

File: pallets/thea/src/lib.rs [hunk 6]
```diff
@@ -229,84 +297,41 @@ pub mod pallet {
...
-			));
+		///Approve Deposit
+		#[pallet::weight(1000)]
+		pub fn approve_deposit(
+			origin: OriginFor<T>,
+			bit_map: u128,
+			bls_signature: [u8; 96],
+			token_type: TokenType,
+			payload: Vec<u8>,
+		) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Please move back parameters definition within the dockblock which removed in this PR for this method.

File: pallets/thea/src/lib.rs [hunk 7]
```diff
@@ -229,84 +297,41 @@ pub mod pallet {
...
-		pub fn approve_deposit(
+		pub fn add_relayer_info(
 			origin: OriginFor<T>,
-			bit_map: u128,
-			bls_signature: [u8; 96],
-			payload: Payload<T::AccountId>,
+			network_id: u8,
+			bls_key: [u8; 192],
+			who: T::AccountId,
 		) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:313]: Please add parameters into the dockblock

File: pallets/thea/src/lib.rs [hunk 8]
```diff
@@ -229,84 +297,41 @@ pub mod pallet {
...
+			let mut current_bls = Self::get_relayers_key_vector(network_id);
+			let mut current_ecdsa = Self::get_auth_list(network_id);
+			let key = BLSPublicKey(bls_key);
 
-			// Update deposit Nonce
-			<DepositNonce<T>>::insert(payload.network_id.saturated_into::<Network>(), nonce + 1);
+			// Update Storage
+			current_bls.try_push(key).unwrap();
+			<RelayersBLSKeyVector<T>>::insert(network_id, current_bls);
+			current_ecdsa.try_push(who.clone()).unwrap();
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Please handle properly results without `unprap`s

File: pallets/thea/src/lib.rs [hunk 9]
```diff
@@ -381,22 +406,11 @@ pub mod pallet {
...
-		/// * `origin`: User
-		/// * `asset_id`: Asset id
-		/// * `amount`: Amount of asset to withdraw
-		/// * `beneficiary`: beneficiary of the withdraw
-		/// * `pay_for_remaining`: user is ready to pay for remaining pending withdrawal for quick
-		///   withdrawal
-		// TODO: [Issue #606] Use benchmarks
+		/// Initiate Withdraw
 		#[pallet::weight(1000)]
 		pub fn withdraw(
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Please move back parameters into the dockblock.

File: pallets/thea/src/lib.rs [hunk 10]
```diff
@@ -408,19 +422,67 @@ pub mod pallet {
...
-			// TODO: This will be refactored when work on withdrawal so not fixing clippy suggestion
-			let network = <AssetIdToNetworkMapping<T>>::get(asset_id)
-				.ok_or_else(|| Error::<T>::UnableFindNetworkForAssetId)?;
+		/// Add Token Config
+		#[pallet::weight(1000)]
+		pub fn set_withdrawal_fee(
+			origin: OriginFor<T>,
+			network_id: u8,
+			fee: u128,
+		) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:465]: Please extend the dockblock with parameters description

File: pallets/thea/src/lib.rs [hunk 11]
```diff
@@ -408,19 +422,67 @@ pub mod pallet {
...
+
+		pub fn do_withdraw(
+			user: T::AccountId,
+			asset_id: u128,
+			amount: u128,
+			beneficiary: Vec<u8>,
+			pay_for_remaining: bool,
+		) -> Result<(), DispatchError> {
+			ensure!(beneficiary.len() <= 100, Error::<T>::BeneficiaryTooLong);
+			//#[allow(clippy::unnecessary_lazy_evaluations)]
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs]: Please remove

File: pallets/thea/src/tests.rs [hunk 12]
```diff
@@ -13,92 +13,585 @@
...
 use crate::{
 	mock,
 	mock::{new_test_ext, Test, *},
 	pallet::*,
 };
 use blst::min_sig::*;
-use thea_primitives::BLSPublicKey;
+use frame_support::traits::fungibles::Mutate;
+use sp_runtime::traits::ConstU32;
+// use frame_support::metadata::StorageEntryModifier::Default as ;
```
Comment (serhii-temchenko) [inline pallets/thea/src/tests.rs]: Please remove

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

**Review by serhii-temchenko — APPROVED** [review]

**Review by Gauthamastro — APPROVED** [review]

**Pull request merged**